### PR TITLE
Improve interoperability with OpenSSL

### PR DESCRIPTION
**Pull request by mgerstner**
- a new tool `tpm2_convert_key` to convert the keys written by `tpm2_akparse` into OpenSSL PEM or DER format in a programmatical way
- new parameter `-r` for `tpm2_sign` for optionally writing the plain signature into a file that can be directly used by OpenSSL tools
- new parameter `-r` and `-t` for `tpm2_quote` for optionally writing the plain signature and plain quote message into separate files that can be directly used by OpenSSL tools

### CODE REVIEW

File: tools/tpm2_convert_key.c [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+            );
+
+            goto out;
+        }
+        else if (data->num_blobs == MAX_KEY_BLOBS) {
+            LOG_ERR("Maximum number of key blobs %d exceeded.", MAX_KEY_BLOBS);
+            goto out;
+        }
+
+        blob.data = malloc(blob.length);
```
Comment (williamcroberts) [inline tools/tpm2_convert_key.c]: bolb.data possible null pointer de-reference when malloc fails.

File: tools/tpm2_convert_key.c [hunk 2]
```diff
@@ -0,0 +1,414 @@
+//**********************************************************************;
+// Copyright (c) 2015, Intel Corporation
```
Comment (williamcroberts) [inline tools/tpm2_convert_key.c]: 2017

File: tools/tpm2_convert_key.c [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+
+/*
+ * Read the binary structure from infile and put it into the given blob output
+ * parameter. This expects the structure as output by tpm2_akparse.
+ */
+static bool read_key(const char *infile, struct key_data *data) {
+    FILE *f;
+    bool ret = false;
+
+    memset(data, 0, sizeof(struct key_data));
```
Comment (williamcroberts) [inline tools/tpm2_convert_key.c]: Any reason you just don't initialize this to 0 when it's declared on the stack?

File: tools/tpm2_convert_key.c [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+        break;
+    default:
+        LOG_ERR("Unexpected output format encountered");
+        goto out;
+    }
+
+    ret = true;
+
+out:
+    if (f)
```
Comment (williamcroberts) [inline tools/tpm2_convert_key.c]: brackets single if.

File: tools/tpm2_convert_key.c [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+    FMT_PEM,
+    FMT_DER,
+    FMT_INVAL
+} convert_format;
+
+// reads just the algorithm type identifier from the given file and returns it
+// in the correct endianess. error handling included.
+static bool read_alg_type(FILE *f, UINT16 *alg_type) {
+
+    if (fread(alg_type, 2, 1, f) != 1) {
```
Comment (williamcroberts) [inline tools/tpm2_convert_key.c]: here and throughout, don't hide functions in conditionals, ie this should be:
```
size_t cnt = fread();
if (cnt != 1) {
   ERROR
}
```

File: tools/tpm2_convert_key.c [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+            }
+
+            ret = true;
+            break;
+        }
+
+        // correct byte order
+        blob.length = tpm2_util_ntoh_16(blob.length);
+
+        if (blob.length > MAX_LENGTH) {
```
Comment (williamcroberts) [inline tools/tpm2_convert_key.c]: Why bound this on 8192, since it's bounded on UINT16, just use it?

File: tools/tpm2_convert_key.c [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+    char *key_output_format;
+};
+
+// holds a single key blob as found in the input file
+struct key_blob {
+    UINT16 length;
+    UINT8  *data;
+};
+
+#define MAX_KEY_BLOBS 8
```
Comment (williamcroberts) [inline tools/tpm2_convert_key.c]: Why?

File: tools/tpm2_convert_key.c [hunk 8]
```diff
@@ -0,0 +1,414 @@
...
+    if (argc <=1 || argc > (int) (2 * sizeof(options) / sizeof(struct option))) {
+        showArgMismatch(argv[0]);
+        return false;
+    }
+
+    int opt;
+    while ((opt = getopt_long(argc, argv, "f:k:t:hv", options, NULL)) != -1) {
+        switch (opt) {
+        case 'f':
+            if (!optarg) {
```
Comment (williamcroberts) [inline tools/tpm2_convert_key.c]: optarg is never NULL if it has a required argument, no need to check.

File: tools/tpm2_convert_key.c [hunk 9]
```diff
@@ -0,0 +1,414 @@
...
+            LOG_ERR("?? getopt returned character code 0%o ??", opt);
+            return false;
+        }
+    }
+
+    if (
+        !ctx->key_input_file_path ||
+        !ctx->key_output_file_path ||
+        !ctx->key_output_format) {
+        LOG_ERR("One or more required parameters missing.");
```
Comment (williamcroberts) [inline tools/tpm2_convert_key.c]: indent

File: tools/tpm2_convert_key.c [hunk 10]
```diff
@@ -0,0 +1,414 @@
...
+            LOG_ERR("Unknown Argument: %c", optopt);
+            return false;
+        default:
+            LOG_ERR("?? getopt returned character code 0%o ??", opt);
+            return false;
+        }
+    }
+
+    if (
+        !ctx->key_input_file_path ||
```
Comment (williamcroberts) [inline tools/tpm2_convert_key.c]: this line should be with the if

File: tools/tpm2_convert_key.c [hunk 11]
```diff
@@ -0,0 +1,414 @@
...
+        TSS2_SYS_CONTEXT *sapi_context) {
+
+    /* opts is unused, avoid compiler warning */
+    (void)opts;
+    (void)sapi_context;
+    (void)envp;
+
+    struct tpm_convert_key_ctx ctx;
+
+    memset(&ctx, 0, sizeof(struct tpm_convert_key_ctx));
```
Comment (williamcroberts) [inline tools/tpm2_convert_key.c]: The downside to this is that it also writes the padding bytes (if any)... but I really don't care.

File: man/tpm2_convert_key.8.in [hunk 12]
```diff
@@ -0,0 +1,62 @@
+.\" Copyright (c) 2016, Intel Corporation
```
Comment (williamcroberts) [inline man/tpm2_convert_key.8.in]: 2017

File: lib/tpm2_alg_util.c [hunk 13]
```diff
@@ -300,3 +300,45 @@ bool pcr_parse_digest_list(char **argv, int len,
...
+bool tpm2_extract_plain_signature(UINT8 **buffer, UINT16 *size,
+    TPMT_SIGNATURE *signature) {
+
+    *buffer = NULL;
+    *size = 0;
+
+    switch (signature->sigAlg) {
+    case TPM_ALG_RSASSA:
+        *size = sizeof((*signature).signature.rsassa.sig.t.buffer);
+        *buffer = malloc(*size);
```
Comment (williamcroberts) [inline lib/tpm2_alg_util.c]: validate malloc return here and throughout

Comment (williamcroberts) [reply lib/tpm2_alg_util.c]: Just return buffer instead of the bool.

File: tools/tpm2_sign.c [hunk 14]
```diff
@@ -166,20 +167,33 @@ static bool sign_and_save(tpm_sign_ctx *ctx) {
...
 
-    static const char *optstring = "k:P:g:m:t:s:c:S:";
+    static const char *optstring = "k:P:g:m:t:s:c:S:r:";
     static const struct option long_options[] = {
       {"keyHandle",1,NULL,'k'},
       {"pwdk",1,NULL,'P'},
       {"halg",1,NULL,'g'},
       {"msg",1,NULL,'m'},
       {"sig",1,NULL,'s'},
+      {"plainSig",1,NULL,'r'},
```
Comment (williamcroberts) [inline tools/tpm2_sign.c]: no camel casing... were moving to get rid of all this camel casing non-sense.

File: tools/tpm2_sign.c [hunk 15]
```diff
@@ -262,6 +276,15 @@ static bool init(int argc, char *argv[], tpm_sign_ctx *ctx) {
             flags.s = 1;
         }
             break;
+        case 'r': {
+            bool result = files_does_file_exist(optarg);
```
Comment (williamcroberts) [inline tools/tpm2_sign.c]: I also dislike how the tools complain that a file exists, it should just overwrite it.

File: Makefile.am [hunk 16]
```diff
@@ -84,7 +84,8 @@ bin_PROGRAMS = \
     tools/tpm2_unseal \
     tools/tpm2_dictionarylockout \
     tools/tpm2_createpolicy \
-    tools/tpm2_pcrextend
+    tools/tpm2_pcrextend \
+    tools/tpm2_convert_key
```
Comment (williamcroberts) [inline Makefile.am]: I think we want is an auxiliary tools option that can be configured, but I am not 100% sold on it. IMHO, the tools should be things that use the TPM2.0 spec. Since this is just a helper, I would make an tools/aux/ folder for this and have configure --enable-aux-tools

File: tools/tpm2_quote.c [hunk 17]
```diff
@@ -212,6 +214,81 @@ UINT16  calcSizeofTPMT_SIGNATURE( TPMT_SIGNATURE *sig )
     return size > sizeof(*sig) ? sizeof(*sig) : size;
 }
 
+FILE* open_file(const char *path)
+{
+    FILE *fp = fopen(path,"w+");
+
+    if(NULL == fp)
```
Comment (williamcroberts) [inline tools/tpm2_quote.c]: if NULL I am, aka yoda variables, please change these.
```
if (!fp)
```

File: tools/tpm2_quote.c [hunk 18]
```diff
@@ -212,6 +214,81 @@ UINT16  calcSizeofTPMT_SIGNATURE( TPMT_SIGNATURE *sig )
     return size > sizeof(*sig) ? sizeof(*sig) : size;
 }
 
+FILE* open_file(const char *path)
+{
+    FILE *fp = fopen(path,"w+");
+
```
Comment (williamcroberts) [inline tools/tpm2_quote.c]: no newline  here.

File: tools/tpm2_quote.c [hunk 19]
```diff
@@ -212,6 +214,81 @@ UINT16  calcSizeofTPMT_SIGNATURE( TPMT_SIGNATURE *sig )
     return size > sizeof(*sig) ? sizeof(*sig) : size;
 }
 
+FILE* open_file(const char *path)
+{
```
Comment (williamcroberts) [inline tools/tpm2_quote.c]: Even though not all files comply or even come close to the coding standard, new hunks should comply, please see: misc/coding_standard_c.txt

File: tools/tpm2_quote.c [hunk 20]
```diff
@@ -212,6 +214,81 @@ UINT16  calcSizeofTPMT_SIGNATURE( TPMT_SIGNATURE *sig )
...
+    FILE *fp = fopen(path,"w+");
+
+    if(NULL == fp)
+    {
+        printf("OutFile: %s Can Not Be Created !\n",path);
```
Comment (williamcroberts) [inline tools/tpm2_quote.c]: use the LOG_ macros.

File: tools/tpm2_quote.c [hunk 21]
```diff
@@ -212,6 +214,81 @@ UINT16  calcSizeofTPMT_SIGNATURE( TPMT_SIGNATURE *sig )
...
+    }
+
+    return fp;
+}
+
+bool write_file(FILE *fp, void *data, size_t len,
```
Comment (williamcroberts) [inline tools/tpm2_quote.c]: files.h has routines for these

File: tools/tpm2_quote.c [hunk 22]
```diff
@@ -305,6 +362,8 @@ int execute_tool (int argc, char *argv[], char *envp[], common_opts_t *opts,
         {"outFile",1,NULL,'o'},
         {"qualifyData",1,NULL,'q'},
         {"input-session-handle",1,NULL,'S'},
+        {"plainSig",1,NULL,'r'},
+        {"plainQuote",1,NULL,'t'},
```
Comment (williamcroberts) [inline tools/tpm2_quote.c]: no camel casing, here and throughout.

File: tools/tpm2_quote.c [hunk 23]
```diff
@@ -390,6 +451,29 @@ int execute_tool (int argc, char *argv[], char *envp[], common_opts_t *opts,
             }
             o_flag = 1;
             break;
+        case 'r': {
+            outPlainSigFilePath = optarg;
+            if(files_does_file_exist(optarg))
```
Comment (williamcroberts) [inline tools/tpm2_quote.c]: drop these file exist checks... here and throughout.

File: tools/tpm2_quote.c [hunk 24]
```diff
@@ -406,13 +490,13 @@ int execute_tool (int argc, char *argv[], char *envp[], common_opts_t *opts,
              }
              is_auth_session = true;
              break;
-       case ':':
+        case ':':
```
Comment (williamcroberts) [inline tools/tpm2_quote.c]: unrelated change, separate patch for this and line 499.

File: tools/tpm2_convert_key.c (hunk 6 above)
Comment (mgerstner) [reply tools/tpm2_convert_key.c]: For being explicit and conservative. Typical public key sized should fit into this for a while.

File: tools/tpm2_convert_key.c (hunk 1 above)
Comment (mgerstner) [reply tools/tpm2_convert_key.c]: is handled now

File: tools/tpm2_convert_key.c (hunk 3 above)
Comment (mgerstner) [reply tools/tpm2_convert_key.c]: Because then the function would rely on how it's called instead of having defined behaviour by design. This way even if a caller fails to initialize the buffer correctly and to interpret the function success indication we'll have a defined execution flow.

File: lib/tpm2_alg_util.c (hunk 13 above)
Comment (mgerstner) [reply lib/tpm2_alg_util.c]: I've adjusted this accordingly.

File: Makefile.am (hunk 16 above)
Comment (mgerstner) [reply Makefile.am]: I don't know if this justifies the effort given that this new tool doesn't introduce any new dependencies or anything. Should a lot of these not stricly spec related tools accumulate it is something that could still be done then.

File: tools/tpm2_quote.c (hunk 21 above)
Comment (mgerstner) [reply tools/tpm2_quote.c]: changed the call inside write_file(). Kept the function, because of the added error output handling present there.

File: tools/tpm2_quote.c (hunk 24 above)
Comment (mgerstner) [reply tools/tpm2_quote.c]: I can do this as part of a general rebase once the rest is conforming.

**Comment by mgerstner** [comment]
Regarding some of the remaining not yet addressed comments in `tpm_quote.c` and `tpm_sign.c` like:

- Yoda checks
- file existence checks
- camel case parameters
- coding style
- `printf` instead of `LOG_`

I did write the changes in the style found in the source files they're contained in. I don't think it makes sense to introduce inconsistency now for the sake of adding a few coding style consistent lines (will be even more confusing). I also think it's a bit much to charge me with refactoring existing source files completely just for adding a couple of lines there.

IMHO such refactoring should be handled in separate commits and PRs. Knowing of these shortcomings now I might be able to address them in the future as time allows.

**Comment by williamcroberts** [comment]
@mgerstner Nobody asked you to change the whole file, just your hunks. The whole file at some point in time needs to be clean up, ill take this into consideration in the PR (ie don't worry about formatting your hunks correctly). With that said, not sure if we really want a tool to do this.

However, using LOG_ is correct, and we don't want to add anymore printfs to the code base. So functional changes will need to be made.

File: Makefile.am (hunk 16 above)
Comment (williamcroberts) [reply Makefile.am]: We already tpm2_akparse. We can revisit this later.

**Comment by williamcroberts** [comment]
Why add a new tool, why can't this functionality be in tpm2_akparse?

**Comment by liuqun** [comment]
It would be wonderful if tpm2_readpublic can save an RSA pubkey as OpenSSL PEM/DER format and tpm2_loadexternal can read external PEM/DER format RSA pubkey into TPM!

**Comment by mgerstner** [comment]
@williamcroberts I can adjust my hunks to the official coding style. What I meant is that right now you have consistently inconsistent source files, and when adding new hunks like this you will have inconsistently inconsistent source files.

Since it is not quite clear to me what the actual purpose of the `tpm2_akparse` was or is I didn't feel comfortable extending it this way. Also its name which is seemingly refering to the AK is somewhat confusing so having an explicit tool for conversion seemed a clearer approach to me. Of course, having the dependency on the `tpm2_akparse` format, the new tool is not completely freed of this confusion (which may only be in my mind, I don't know).

Maybe what @liuqun suggests, adding the functionality to `tpm2_readpublic` is the right way to go. Having a tool dedicated to format conversion has it's advantages, too. It's immediately clear from the name so when you're looking for format conversion you will quickly find the right starting spot. It follows the _do one thing and do it well_ idiom which I personally like. Also since you seem to be concerned about what is actual TPM spec related and what is auxiliary, a separate tool for converions provides a better separation of concerns.

Well I will not continue adjusting this PR if you're not convinced of the feature or the approach taken. Let me know what you'd like.

**Comment by williamcroberts** [comment]
@mgerstner "** consistently inconsistent source files, and when adding new hunks like this you will have inconsistently inconsistent source files** -- Totally agree. I ended up taking over this project in quite a messy state, and am trying to work toward unifying it in time.

As far as the TPM spec, I just want tools that somewhat line up, but are also easy to use (unlike the tpm and its spec :-1): . tpm2_readpublic, should at some point send that command to the tpm. (yes tpm2_getmanufec and likely others detour from this).

I think being able to control output format of outputted data is better than having to build up something, if your already calling tpm2_readpublic and you just want the public key in pem/der format it would be nice to just specify it.
```
tpm2_readpublic --format=pem
```
is more succinct than:
tpm2_readpublic | tpm2_convert --inform=fromreadpublic --outform=pem

**Comment by martinezjavier** [comment]
I haven't been following this thread closely, but I just wanted to give my opinion on coding style and consistency. I think that consistency as a project is much more important than consistency in individual source files.

@mgerstner BTW, awesome work! It will be great to have DER/PEM support in tpm2-tools :smiley:

**Comment by mgerstner** [comment]
@williamcroberts Yes the approach with `tpm2_readpublic --format=pem` is nice. However I encountered this with `tpm2_getpubak -f public.key ...`. Other commands like `tpm2_getpubek` and `tpm2_create` also have parameters that write the public key to disk.

The single `tpm2_convert_key` would leave the existing commands untouched, users could use the TPM specific format and convert it explicitly, if desired.

If you prefer the `--format=...` approach then I'll have to open a new pull request. It will require some more time to add the new behaviour to all existing commands that are candidates for this. Or the existing commands will not get the new behaviour and users always have to go through `tpm2_readpublic` for getting PEM/DER format keys, which would save me some time.

**Comment by williamcroberts** [comment]
Obviously I would love a PR with all tools updated, but I would accepts just one tool, with a ticket opened with the others that need to be modified.

Right now, were about to have some very disruptive changes on master, as the tool option parsing code, was very, very broken.

We ended up pulling options up into a single getopt_long parser to correct the behavior.

More info here:
https://github.com/01org/tpm2-tools/pull/446

We'll be working on tpm2-option-revamp branch, ill rectify the option handling with stuff on master, but I am trying to avoid putting big changes on master to make it easy to rectify.

**Comment by mgerstner** [comment]
@williamcroberts so to summarize this: you'd like to go for the `--format=pem|der|...` approach at best implemented in all tools that output public keys in some way. But at the moment disruptive changes to the option handling are in the works.

How long do you think this will take to be finished? You could give me a hint when it's finished so I can make another attempt on this. I'd hate to deal with a bunch of merge conflicts...

File: tools/tpm2_quote.c [hunk 25]
```diff
@@ -212,6 +214,84 @@ UINT16  calcSizeofTPMT_SIGNATURE( TPMT_SIGNATURE *sig )
     return size > sizeof(*sig) ? sizeof(*sig) : size;
 }
 
+FILE* open_file(const char *path)
+{
+    FILE *fp = fopen(path,"w+");
```
[the remark this answers is not in the document]
Comment (mgerstner) [reply tools/tpm2_quote.c]: note taken. since this PR won't make it through I'll consider it for the next approach.

**Comment by martinezjavier** [comment]
@mgerstner about your summary, that's what he meant IIUC.

It shouldn't take long, we already agreed on the core changes so it's just a matter of porting the tools to the new infrastructure. I'll help @williamcroberts with tha and I expect to dedicate some time on it later today.

**Comment by williamcroberts** [comment]
@mgerstner I am hoping to land this disruptive change on master by Sept 8

**Comment by williamcroberts** [comment]
@mgerstner I have landed the changes on master, so if you want to re-send a PR please do.

I'm going to close this one, and you can re-submit. Thank you.
